bigtable/cmd/cbt/cbt_test.go

```diff
@@ -171,3 +178,401 @@ func TestParseColumnsFilter(t *testing.T) {
+				continue
+			}
+		}
+		for i, v := range line {
+			if v != tc.nextLine[i] {
+				t.Errorf("parseImportArgs() did not fail for next line idx:%d, got: %q, want %q", i, cols[i], tc.oCols[i])
+				continue
+			}
+		}
+	}
+}
+
+func setupEmulator(t *testing.T, tables, families []string) (context.Context, *bigtable.Client) {
+	srv, err := bttest.NewServer("localhost:0")
+	if err != nil {
+		t.Fatalf("Error starting bttest server: %s", err)
+	}
+
+	ctx := context.Background()
+
+	conn, err := grpc.Dial(srv.Addr, grpc.WithInsecure())
+	if err != nil {
+		t.Fatalf("Error %s", err)
+	}
+
+	proj, instance := "proj", "instance"
+	adminClient, err := bigtable.NewAdminClient(ctx, proj, instance, option.WithGRPCConn(conn))
+	if err != nil {
+		t.Fatalf("Error %s", err)
+	}
+
+	for _, ta := range tables {
+		if err = adminClient.CreateTable(ctx, ta); err != nil {
+			t.Fatalf("Error %s", err)
+		}
+		for _, f := range families {
+			if err = adminClient.CreateColumnFamily(ctx, ta, f); err != nil {
+				t.Fatalf("Error %s", err)
+			}
+		}
+	}
+
+	client, err := bigtable.NewClient(ctx, proj, instance, option.WithGRPCConn(conn))
+	if err != nil {
+		t.Fatalf("Error %s", err)
+	}
+
+	return ctx, client
+}
+
+func validateData(ctx context.Context, tbl *bigtable.Table, fams, cols []string, rowData [][]string) error {
+	// vaildate table entries, valMap["rowkey:family:column"] = mutation value
+	valMap := make(map[string]string)
+	for _, row := range rowData {
+		for i, val := range row {
+			if i > 0 && val != "" {
+				valMap[row[0]+":"+fams[i]+":"+cols[i]] = val
+			}
+		}
+	}
+	for _, data := range rowData {
+		row, err := tbl.ReadRow(ctx, data[0])
+		if err != nil {
+			return err
+		}
+		for _, cf := range row {
+			for _, column := range cf {
+				k := data[0] + ":" + string(column.Column)
+				v, ok := valMap[k]
+				if ok && v == string(column.Value) {
+					delete(valMap, k)
+					continue
+				}
+			}
+		}
+	}
+	if len(valMap) != 0 {
+		return fmt.Errorf("Data didn't match after read, not found %v", valMap)
+	}
+	return nil
+}
+
+func TestCsvParseAndWrite(t *testing.T) {
+	ctx, client := setupEmulator(t, []string{"my-table"}, []string{"my-family", "my-family-2"})
+
+	tbl := client.Open("my-table")
+	fams := []string{"", "my-family", "my-family-2"}
+	cols := []string{"", "col-1", "col-2"}
+	rowData := [][]string{
+		{"rk-0", "A", "B"},
+		{"rk-1", "", "C"},
+	}
+
+	byteData, err := writeAsCSV(rowData)
+	if err != nil {
+		t.Fatal(err)
+	}
+	reader := csv.NewReader(bytes.NewReader(byteData))
+
+	sr := safeReader{r: reader}
+	if err = sr.parseAndWrite(ctx, tbl, fams, cols, 1, 1, 1); err != nil {
+		t.Fatalf("parseAndWrite() failed unexpectedly")
+	}
+
+	if err := validateData(ctx, tbl, fams, cols, rowData); err != nil {
+		t.Fatalf("Read back validation error:%s", err)
+	}
+}
+
+func TestCsvParseAndWriteBadFamily(t *testing.T) {
+	ctx, client := setupEmulator(t, []string{"my-table"}, []string{"my-family"})
```

Comment:
I wonder if this couldn't use the NewEmulateEnv? 
https://github.com/googleapis/google-cloud-go/blob/6be4cae4555c128f1dee7d778f0b8358d20c6cee/bigtable/export_test.go#L150


Related, but not likely relevant to these tests: 

Also, we have some tests that don't work in emulator or prod. 

```
if !testEnv.Config().UseProd {
		t.Skip("emulator doesn't support IAM Policy creation")
	}
```

Unless these tests wouldn't work against prod, in which case it may be relevant in reverse?

Reply:
I don't touch anything admin based, so fine on the IAM front. Looking into the emulator now.

Reply:
:+1: sorry for the confusion, I was just showing that our testEnv bits should be able to detect if it was requested we use prod for the test run or not. If you know tests work in emulator and/or prod you can guard them :)

Reply:
For NewEmulatedEnv I think many of the steps in setupEmulator() will still be required as seen here https://github.com/googleapis/google-cloud-go/blob/6be4cae4555c128f1dee7d778f0b8358d20c6cee/bigtable/bigtable_test.go#L181

I tried to replace the bttest.NewServer() call with it and for some reason NewEmulatedEnv isn't being discovered.

Reply:
We are ok running the cbt commands against the emulator as the behaviour should be as expected. It could be another feature to add integration test support for prod, however the current setup has this tucked in integration_test.go and exporter_test.go files in the parent package, and setting up compatibility with downstream cbt methods (cyclic imports) and is outside scope currently.